keycache/keycache.go

```diff
@@ -45,12 +61,50 @@ func setUser(in ActiveUser, name string) {
 	UserKeys[name] = in
 }
 
+// matches returns true if this usage applies the user and label
+// an empty array of Users indicates that all users are valid
+func (usage Usage) matchesLabel(labels []string) bool {
+	// if asset has no labels always match
+	if len(labels) == 0 {
+		return true
+	}
+	//
+	for _, validLabel := range usage.Labels {
+		for _, label := range labels {
+			if label == validLabel {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// matches returns true if this usage applies the user and label
+// an empty array of Users indicates that all users are valid
+func (usage Usage) matches(user string, labels []string) bool {
+	if !usage.matchesLabel(labels) {
+		return false
+	}
+	// if usage lists no users, always match
+	if len(usage.Users) == 0 {
+		return true
+	}
+	for _, validUser := range usage.Users {
+		if user == validUser {
+			return true
+		}
+	}
+	return false
+}
+
 // useKey decrements the counter on an active key
 // for decryption or symmetric encryption
-func useKey(name string) {
-	if val, present := matchUser(name); present {
-		val.Uses -= 1
+func useKey(name, user string, labels []string) {
+	if val, present := matchUser(name, user, labels); present {
+		val.Usage.Uses -= 1
 		setUser(val, name)
+	} else {
+		log.Println("Can't use key, not matched", name, user, labels)
```

Comment:
Log text is a little ambigious; a user who hasn't looked at the code might not understand the context of "not matched". Maybe something along the lines of "Key is unusable: an appropriate user could not be found."